plugins/lookup/reverse_lookup.py

```diff
@@ -0,0 +1,175 @@
+        description:
+            - How often to retry on SERVFAIL errors.
+        type: int
+        default: 0
+notes:
+    - Note that when using this lookup plugin with V(lookup(\)), and the result is a one-element list,
+      Ansible simply returns the one element not as a list. Since this behavior is surprising and
+      can cause problems, it is better to use V(query(\)) instead of V(lookup(\)). See the examples
+      and also R(Forcing lookups to return lists, query) in the Ansible documentation.
+'''
+
+EXAMPLES = """
+- name: Look up hostname of IPv4 address
+  ansible.builtin.debug:
+    msg: "{{ query('community.dns.reverse_lookup', '192.168.1.1') }}"
+
+- name: Look up hostname of IPv6 address
+  ansible.builtin.debug:
+    msg: "{{ query('community.dns.reverse_lookup', '1:2:3::4') }}"
+"""
+
+RETURN = """
+_result:
+    description:
+        - The hostname(s) returned for the queried IP addresses.
+        - If multiple IP addresses are queried in O(_terms), the resulting lists have been concatenated.
+    type: list
+    elements: str
+    sample:
+        - example.com
+        - example.org
+"""
+
+from ansible.errors import AnsibleLookupError
+from ansible.plugins.lookup import LookupBase
+from ansible.module_utils.common.text.converters import to_text
+
+from ansible_collections.community.dns.plugins.module_utils.ips import (
+    is_ip_address,
+)
+
+from ansible_collections.community.dns.plugins.module_utils.resolver import (
+    SimpleResolver,
+)
+
+from ansible_collections.community.dns.plugins.plugin_utils.ips import (
+    assert_requirements_present as assert_requirements_present_ipaddress,
+)
+
+from ansible_collections.community.dns.plugins.plugin_utils.resolver import (
+    assert_requirements_present as assert_requirements_present_dnspython,
+    guarded_run,
+)
+
+try:
+    import dns.resolver
+except ImportError:
+    # handled by assert_requirements_present_dnspython
+    pass
+
+try:
+    import ipaddress
+except ImportError:
+    # handled by assert_requirements_present
+    pass
+
+
+class LookupModule(LookupBase):
+    @staticmethod
+    def _resolve(resolver, name, rdtype, server_addresses):
+        def callback():
+            try:
+                rrset = resolver.resolve(
+                    name,
+                    rdtype=rdtype,
+                    server_addresses=server_addresses,
+                    nxdomain_is_empty=True,
+                    target_can_be_relative=False,
+                    search=False,
+                )
+                if not rrset:
+                    return []
+                return [to_text(data) for data in rrset]
+            except dns.resolver.NXDOMAIN:
+                raise AnsibleLookupError('Got NXDOMAIN when querying {name}'.format(name=name))
+
+        return guarded_run(
+            callback,
+            error_class=AnsibleLookupError,
+            server=name,
+        )
+
+    def run(self, terms, variables=None, **kwargs):
```

Comment:
```suggestion
    def run(self, terms: List[str], variables: Optional[dict] = None, **kwargs: Any): -> List[str]:
```

Reply:
Instead of `Optional[dict]` you can also use `dict | None`. That requires either 3.9 or 3.10 (I'm not sure anymore), but with the right `__future__` import (I think `from __future__ import annotations`) you can already start using it in Python 3.7+ (as long as the type checking happens with a new enough Python).